server.go

```diff
@@ -160,6 +171,500 @@ func (s *server) RemoveRebroadcastInventory(iv *wire.InvVect) {
 	s.modifyRebroadcastInv <- broadcastInventoryDel(iv)
 }
 
+// handleVersionMsg returns a handler which is invoked when a peer
+// receives a version bitcoin message and is used to negotiate the
+// protocol version details as well as kick start the communications.
+func (s *server) handleVersionMsg(p *peer.Peer, msg *wire.MsgVersion) {
+	// Add the remote peer time as a sample for creating an offset against
+	// the local clock to keep the network time in sync.
+	s.timeSource.AddTimeSample(p.Addr(), msg.Timestamp)
+
+	// Signal the block manager this peer is a new sync candidate.
+	s.blockManager.NewPeer(p)
+	s.blockManager.WaitForPeer()
+
+	// Choose whether or not to relay transactions before a filter command
+	// is received.
+	pd := s.blockManager.peers[p.ID()]
+	pd.relayMtx.Lock()
+	defer pd.relayMtx.Unlock()
```

Comment:
I'd not defer the unlock here since there is only one exist path and it's much cheaper to avoid the defer.